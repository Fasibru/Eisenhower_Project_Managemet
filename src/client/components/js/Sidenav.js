/*
This component serves as the section for setting filters to slice and dice the data/tasks
Filtered tasks will then be rendered in the Main component.
*/

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import '../css/Sidenav.css';
// import { NewTask } from './NewTask'

class Sidenav extends Component {
  /*
  - implement filters
  - filters should update the task array and set it as state in the App
  */

  logTasks = () => {
    const { tasks } = this.props;
    console.log('All Tasks:', tasks);
  }

  render() {
    const {
      toggleNewTaskPopup,
      handleFilterCompleted,
      filters,
    } = this.props;
    return (
      <aside className="sidenav">
        <ul className="sidenav__list">
          <li className="sidenav__list-elem">Search field</li>
          <li className="sidenav__list-elem">
            <button type="button" onClick={toggleNewTaskPopup}>New Task</button>
          </li>
          <li className="sidenav__list-elem">
            {/* <button type="button" onClick={this.logTasks}>Log Tasks</button> */}
            <p>Show tasks:</p>
            <form>
              <select name="completed" value={filters.showCompleted} onChange={handleFilterCompleted}>
                <option value="both">Both</option>
                <option value="false">Open</option>
                <option value="true">Completed</option>
              </select>
            </form>
          </li>
          <li className="sidenav__list-elem">Specify Filters (filter section below)</li>
        </ul>
      </aside>
    );
  }
}

Sidenav.defaultProps = {
  filters: {
    showCompleted: 'yes',
  },
};

Sidenav.propTypes = {
  tasks: PropTypes.arrayOf(PropTypes.object).isRequired,
  toggleNewTaskPopup: PropTypes.func.isRequired,
  handleFilterCompleted: PropTypes.func.isRequired,
  filters: PropTypes.shape({
    showCompleted: PropTypes.string,
  }),
};

export default Sidenav;
