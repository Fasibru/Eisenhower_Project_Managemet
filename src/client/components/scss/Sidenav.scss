@import './variables';

.sidenav {
  // display: flex;
  // flex-direction: column;
  grid-area: sidenav;
  background-color: $component-color-sidenav;
  // position: sticky;
  // top: 0;
  ul {
    margin-top: 0;
    position: sticky;
    top: 0;
  }
}

.sidenav__list {
  display: flex;
  flex-direction: column;
  // margin-top: 50px;
  padding-left: 20px;
  padding-right: 20px;
  list-style-type: none;
}

.sidenav__list-elem {
  color: $element-color;
  padding: 20px 0px 20px 0px;
  flex: 1;
  display: flex;
  flex-wrap: wrap;

  p {
    margin: 0;
  }
}

.sidenav__search {
  flex: 1;
}

.sidenav__status {
  flex: 1;
}

.sidenav__date {
  display: flex;
  flex-direction: column;
  flex: 1;

  input {
    font-family: $font-stack;
    background: $element-color;
  }
}


.sidenav__showTasks {
  flex: 1;
  display: flex;
}

.sidenav__btn {
  flex: 1;
  background: $element-color;
  border: none;
  padding: 10px;
  transition-duration: 0.1s;
  -webkit-appearance: none;
}

.sidenav__btn:hover {
  background: $hover-color;
}

.sidenav__btn:focus {
  outline-style: none;
  background: $focus-color;
}

.sidenav__btn--border-left-right {
  border-left: 0.5px solid $hover-color;
  border-right: 0.5px solid $hover-color;
}

.sidenav__btn--active {
  background: $hover-color;
}