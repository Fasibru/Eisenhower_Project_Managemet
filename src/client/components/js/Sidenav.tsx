// import PropTypes from 'prop-types';
import * as React from 'react';
import { connect } from 'react-redux';

import { faTimes } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { openNewTaskPopup } from '../../actions/actionsTasks';
import '../scss/Sidenav.scss';

import DateFilterContainer from '../../containers/DateFilter.container';
import SearchFilter from './SearchFilter';
import StatusFilter from './StatusFilter';

import { TaskActionsTypes } from '../../../types/taskActionTypes';

interface SidenavProps {
  isOpen: boolean;
  toggleFilterMenu(): void;
  openNewTaskPopup(): TaskActionsTypes;
}

// tslint:disable-next-line: variable-name
export const Sidenav: React.FC<SidenavProps> = ({
  // tslint:disable-next-line: no-shadowed-variable
  openNewTaskPopup,
  isOpen,
  toggleFilterMenu,
}) => {
  return (
    <aside
      className={isOpen ? 'sidenav sidenav--active' : 'sidenav'}
    >
      <button
        type="button"
        className="sidenav__close-icon"
        onClick={toggleFilterMenu}
        onMouseDown={e => e.preventDefault()} /* to remove focus after button is clicked */
      >
        <FontAwesomeIcon icon={faTimes} />
      </button>
      <ul className="sidenav__list">
        <li className="sidenav__list-elem">
          <button
            className="sidenav__btn"
            type="button"
            onClick={openNewTaskPopup}
          >
            Add New Task
          </button>
        </li>
        <li className="sidenav__list-elem">
          <SearchFilter />
        </li>
        <li className="sidenav__list-elem">
          <StatusFilter />
        </li>
        <li className="sidenav__list-elem">
          <DateFilterContainer />
        </li>
        <li
          className="sidenav__list-elem"
        >
          Specify further filters e.g. by team member (very much in the future)
        </li>
      </ul>
    </aside>
  );
};

// Sidenav.propTypes = {
//   openNewTaskPopup: PropTypes.func.isRequired,
//   toggleFilterMenu: PropTypes.func.isRequired,
//   isOpen: PropTypes.bool.isRequired,
// };

export default connect(null, {
  openNewTaskPopup,
})(Sidenav);
