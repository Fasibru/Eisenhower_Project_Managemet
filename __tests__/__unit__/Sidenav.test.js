import React from 'react';
import { shallow } from 'enzyme';
import Sidenav from '../../src/client/components/js/Sidenav';

describe('Testing <Sidenav />', () => {
  const fnClick = jest.fn();
  const fnChange = jest.fn();
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Sidenav
      toggleNewTaskPopup={fnClick}
      filters={{
        showTasks: 'both',
      }}
      handleFilterShowTasks={fnChange}
    />);
  });

  it('Renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should populate "select" based on provided props', () => {
    const select = wrapper.find('select');

    expect(select.props().value).toEqual('both');
  });

  it('Should call "handleFilterShowTasks" on change', () => {
    const event = {
      target:
      { value: '' },
    };
    const select = wrapper.find('select');
    select.simulate('change', event);

    expect(fnChange).toHaveBeenCalledTimes(1);
  });

  it('Should call "toggleNewTaskPopup" on button click', () => {
    const button = wrapper.find('button');
    button.simulate('click');

    expect(fnClick).toHaveBeenCalledTimes(1);
  });
});
