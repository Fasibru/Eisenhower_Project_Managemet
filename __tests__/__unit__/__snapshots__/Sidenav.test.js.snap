// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Sidenav Renders correctly 1`] = `
<aside
  className="sidenav"
>
  <ul
    className="sidenav__list"
  >
    <li
      className="sidenav__list-elem"
    >
      Search field
    </li>
    <li
      className="sidenav__list-elem"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        New Task
      </button>
    </li>
    <li
      className="sidenav__list-elem"
    >
      <p>
        Show tasks:
      </p>
      <form>
        <select
          name="showTasks"
          onChange={[Function]}
          value="both"
        >
          <option
            value="both"
          >
            Both
          </option>
          <option
            value="open"
          >
            Open
          </option>
          <option
            value="completed"
          >
            Completed
          </option>
        </select>
      </form>
    </li>
    <li
      className="sidenav__list-elem"
    >
      Specify Filters (filter section below)
    </li>
  </ul>
</aside>
`;
